### PR TITLE
S0074 masked alphanumeric mod97 10

### PR DESCRIPTION
**Pull request by KnowledgeForwardSolutions**
This pull request introduces significant improvements and expanded test coverage for the `AlphanumericMod97_10Algorithm`, focusing on enhanced masking support, algorithm optimization, and more comprehensive unit and benchmark tests. The changes ensure the algorithm is robust when handling masked input (such as values with spaces or dashes), optimize internal calculations, and add thorough validation for both standard and edge cases.

**Key changes include:**

### Masking Support and Algorithm Enhancements

* `AlphanumericMod97_10Algorithm` now implements the `IMaskedCheckDigitAlgorithm` interface, enabling validation of values with non-check digit characters using an `ICheckDigitMask`. The documentation and implementation are updated to clarify and support this new capability.
* Internal optimizations: The calculation of the reduce threshold is now more precise to prevent integer overflow, and letter-to-digit mapping is further optimized by precomputing digit values.

### Expanded and Improved Unit Tests

* Extensive new unit tests added to `AlphanumericMod97_10AlgorithmTests` to cover masked validation scenarios, edge cases (such as null, empty, or insufficient input), correct and incorrect check digit handling, and error detection with various masks. This includes tests for both valid and invalid inputs with different masking strategies. [[1]](diffhunk://#diff-a529948c2d4f0df8cd7635b07959d258ea9b364d55b37525661a787c7cb58560R6-R8) [[2]](diffhunk://#diff-a529948c2d4f0df8cd7635b07959d258ea9b364d55b37525661a787c7cb58560R155-R165) [[3]](diffhunk://#diff-a529948c2d4f0df8cd7635b07959d258ea9b364d55b37525661a787c7cb58560R183) [[4]](diffhunk://#diff-a529948c2d4f0df8cd7635b07959d258ea9b364d55b37525661a787c7cb58560L235-R250) [[5]](diffhunk://#diff-a529948c2d4f0df8cd7635b07959d258ea9b364d55b37525661a787c7cb58560L244-R259) [[6]](diffhunk://#diff-a529948c2d4f0df8cd7635b07959d258ea9b364d55b37525661a787c7cb58560R293-R474)
* Added property tests in `MaskedAlgorithmsTests` to ensure the `MaskedAlgorithms.AlphanumericMod97_10` property is correctly implemented and of the expected type.

### Benchmark and Test Infrastructure Updates

* Updated `OptimizeBenchmarks` to benchmark the new `AlphanumericMod97_10Algorithm` using more complex alphanumeric values, replacing the previous Damm algorithm benchmarks.
* Modified the benchmark runner in `Program.cs` to run the updated `OptimizeBenchmarks` by default. [[1]](diffhunk://#diff-7687108d65e57c8ece664486c28509f07a7fe5f4617870bbbb743f3455783c9bL2-R2) [[2]](diffhunk://#diff-7687108d65e57c8ece664486c28509f07a7fe5f4617870bbbb743f3455783c9bL12-R12)

### Code Quality and Documentation

* Improved code comments and documentation for clarity, especially around how letters are mapped to digits and how masking is handled. [[1]](diffhunk://#diff-c1838ca45d1205fd06b9169d06dd6b09a4f13bb2aa9c1ac58c6013b2250d678dR20-R46) [[2]](diffhunk://#diff-c1838ca45d1205fd06b9169d06dd6b09a4f13bb2aa9c1ac58c6013b2250d678dL108-R138)
* Minor corrections to test method names for consistency. [[1]](diffhunk://#diff-a529948c2d4f0df8cd7635b07959d258ea9b364d55b37525661a787c7cb58560L235-R250) [[2]](diffhunk://#diff-a529948c2d4f0df8cd7635b07959d258ea9b364d55b37525661a787c7cb58560L244-R259)

These changes collectively make the algorithm more flexible, robust, and well-tested, especially for real-world scenarios where input may be formatted for human readability.

---

**References:**  
[[1]](diffhunk://#diff-c1838ca45d1205fd06b9169d06dd6b09a4f13bb2aa9c1ac58c6013b2250d678dR20-R46) [[2]](diffhunk://#diff-c1838ca45d1205fd06b9169d06dd6b09a4f13bb2aa9c1ac58c6013b2250d678dL57-R78) [[3]](diffhunk://#diff-c1838ca45d1205fd06b9169d06dd6b09a4f13bb2aa9c1ac58c6013b2250d678dL108-R138) [[4]](diffhunk://#diff-a529948c2d4f0df8cd7635b07959d258ea9b364d55b37525661a787c7cb58560R6-R8) [[5]](diffhunk://#diff-a529948c2d4f0df8cd7635b07959d258ea9b364d55b37525661a787c7cb58560R155-R165) [[6]](diffhunk://#diff-a529948c2d4f0df8cd7635b07959d258ea9b364d55b37525661a787c7cb58560R183) [[7]](diffhunk://#diff-a529948c2d4f0df8cd7635b07959d258ea9b364d55b37525661a787c7cb58560L235-R250) [[8]](diffhunk://#diff-a529948c2d4f0df8cd7635b07959d258ea9b364d55b37525661a787c7cb58560L244-R259) [[9]](diffhunk://#diff-a529948c2d4f0df8cd7635b07959d258ea9b364d55b37525661a787c7cb58560R293-R474) [[10]](diffhunk://#diff-4dd096f96ce85b178672dab194e55dba9b01c61c59201fc2662072de29a1837aR7-R20) [[11]](diffhunk://#diff-bcfbf601d72d5aba080771f01eb6949556d3dd51b2870219dae9f3f8b51be450L10-R30) [[12]](diffhunk://#diff-7687108d65e57c8ece664486c28509f07a7fe5f4617870bbbb743f3455783c9bL2-R2) [[13]](diffhunk://#diff-7687108d65e57c8ece664486c28509f07a7fe5f4617870bbbb743f3455783c9bL12-R12)

### CODE REVIEW

File: CheckDigits.Net/GeneralAlgorithms/AlphanumericMod97_10Algorithm.cs [hunk 1]
```diff
@@ -105,16 +118,24 @@ public Boolean Validate(String value)
...
          var ch = value[index];
-         if (ch >= Chars.DigitZero && ch <= Chars.DigitNine)
+         if (IsDigitCharacter(ch))
          {
             sum = (sum + ch.ToIntegerDigit()) * _radix;
          }
-         else if ((ch >= Chars.UpperCaseA && ch <= Chars.UpperCaseZ) 
-            || (ch >= Chars.LowerCaseA && ch <= Chars.LowerCaseZ))
+         else if (IsLetterCharacter(ch))
          {
```
[the remark this answers is not in the document]
Comment (KnowledgeForwardSolutions) [reply CheckDigits.Net/GeneralAlgorithms/AlphanumericMod97_10Algorithm.cs:132]: suggestion implemented

File: CheckDigits.Net.Tests.Unit/GeneralAlgorithms/AlphanumericMod97_10AlgorithmTests.cs [hunk 2]
```diff
@@ -275,5 +290,187 @@ public void AlphanumericMod97_10Algorithm_Validate_ShouldReturnFalse_WhenValueCo
...
+      => _sut.Validate(String.Empty, _acceptAllMask).Should().BeFalse();
+
+   [Fact]
+   public void AlphanumericMod97_10Algorithm_ValidateMasked_ShouldReturnFalse_WhenAllNonCheckDigitCharactersAreMaskedOut()
+      => _sut.Validate("U7y 8SX rC0 87", _rejectAllMask).Should().BeFalse();     // Would otherwise be valid if non-check characters were not masked out
+
+   [Theory]
+   [InlineData("A")]
+   [InlineData("AB")]
+   [InlineData("98")]         // Would otherise be true if length requirement was not enforced
```
[the remark this answers is not in the document]
Comment (KnowledgeForwardSolutions) [reply CheckDigits.Net.Tests.Unit/GeneralAlgorithms/AlphanumericMod97_10AlgorithmTests.cs:334]: fixed

File: CheckDigits.Net.Tests.Unit/GeneralAlgorithms/AlphanumericMod97_10AlgorithmTests.cs [hunk 3]
```diff
@@ -275,5 +290,187 @@ public void AlphanumericMod97_10Algorithm_Validate_ShouldReturnFalse_WhenValueCo
...
+   public void AlphanumericMod97Algorithm_ValidateMasked_ShouldReturnFalse_WhenInsufficientUnmaskedCharactersToCalculateCheckDigit(String value)
```
[the remark this answers is not in the document]
Comment (KnowledgeForwardSolutions) [reply CheckDigits.Net.Tests.Unit/GeneralAlgorithms/AlphanumericMod97_10AlgorithmTests.cs:335]: fixed

File: CheckDigits.Net.Tests.Unit/GeneralAlgorithms/AlphanumericMod97_10AlgorithmTests.cs [hunk 4]
```diff
@@ -149,6 +152,17 @@ public void AlphanumericMod97_10Algorithm_TryCalculateCheckDigits_ShouldReturnFa
...
+   [Theory]
+   [InlineData("U7y")]
+   [InlineData("U7y8SX")]
+   [InlineData("U7y8SXrC0")]
+   [InlineData("U7y8SXrC0O3S")]
+   [InlineData("U7y8SXrC0O3Sc4I")]
+   [InlineData("U7y8SXrC0O3Sc4IHYQ")]
+   [InlineData("U7y8SXrC0O3Sc4IHYQF4M")]
+   public void AlphanumericMod97_10Algorithm_TryCalculateCheckDigit_ShouldReturnTrue_ForBenchmarkValues(String value)
+      => _sut.TryCalculateCheckDigits(value, out _, out _).Should().BeTrue();
```
[the remark this answers is not in the document]
Comment (KnowledgeForwardSolutions) [reply CheckDigits.Net.Tests.Unit/GeneralAlgorithms/AlphanumericMod97_10AlgorithmTests.cs:164]: fixed

File: CheckDigits.Net.Tests.Unit/GeneralAlgorithms/AlphanumericMod97_10AlgorithmTests.cs [hunk 5]
```diff
@@ -166,6 +180,7 @@ public void AlphanumericMod97_10Algorithm_Validate_ShouldReturnFalse_WhenInputIs
    [Theory]
    [InlineData("A")]
    [InlineData("AB")]
+   [InlineData("98")]         // Would otherise be true if length requirement was not enforced
```
[the remark this answers is not in the document]
Comment (KnowledgeForwardSolutions) [reply CheckDigits.Net.Tests.Unit/GeneralAlgorithms/AlphanumericMod97_10AlgorithmTests.cs:183]: fixed
